### PR TITLE
CONSOLE-3695: Add details item extension to console dynamic plugin SDK

### PR DESCRIPTION
**Pull request by TheRealJon**
- Add details item extension API
- consume the new extension in the DetailsForKind component
- Address some tech debt

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/details-item.ts
Comment (jerolimov) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/details-item.ts:10]: Can we also pass the `path?: string;` and `value?: any/never` to the component?

1. The path from the extensions would allow us to have one "DateTime" component that could be used for two fields, for example, a "Start time" and "Completion time", etc.
2. Wdyt about extracting and passing also the `value` from `obj`? When rendering the Component you use already `_.get(obj, path, '-')` to extract it when there is no Component. That would allow us to make a reused component even easier. In the example from above, this means Just consuming the DateTime string from the props.

```suggestion
  /** The resource for which details are being rendered. */
  obj: R;

  /**
   * The fully qualified path to a string-type property on the resource as defined in the extension.
   * @example 'spec.template.spec.containers[0].image'
   */
  path?: string;

  /**
   * If the path is defined, the value contains the extracted value from the Kubernetes resource.
   */
  value?: never;
```

File: frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx
Comment (jerolimov) [inline frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx:14]: When `_.get(obj, path)` returns `null`, this will throw an error. The easiest way is maybe just to use `_.get(obj, path) ?? '-'` ?? But for an object, it would still render "[object Object]".

I don't know if it is too much for here, but maybe we can do something like:

1. typeof value === "string" => just render the string value
2. value === null || value === undefined => "-"
3. else -> use JSON.stringify()? // this should work for booleans, numbers, arrays and objects right?

Wdyt?

**Review by jerolimov — COMMENTED** [review]
Great extension! 😄

Two small ideas below. Wdyt?

**Comment by TheRealJon** [comment]
QE Approver
/assign @yufchang

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

**Review by rhamilto — COMMENTED** [review]
Nice work, @TheRealJon.  One nit related to the output of docs.  Will hold on tagging pending resolution of Christoph's comments.

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md
Comment (opayne1) [inline frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:962]: Technically this is passive voice. You could do something like:
```suggestion
| `component` | `CodeRef<React.ComponentType<DetailsItemComponentProps<K8sResourceCommon>>>` | yes | The component will render as the details item value. This takes precedence over the<br/>`path` property. |
```

**Review by opayne1 — COMMENTED** [review]
One small docs nit from me. I am going to go ahead and apply the label.

File: frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx
Comment (TheRealJon) [reply frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx:14]: I think we would only want to render string values if no component is provided. I'll improve the handling of null here.

**Comment by TheRealJon** [comment]
@opayne1 @rhamilto @jerolimov I believe I've addressed everyone's comments. PTAL when you can!

**Comment by opayne1** [comment]
lgtm from docs!

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/details-item.ts
Comment (jerolimov) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/details-item.ts:12]: Minor: It's only defined if the path value of the extension is defined.

```suggestion
  path?: string;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/details-item.ts
Comment (jerolimov) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/details-item.ts:15]: Same:

```suggestion
  value?: V;
```

File: frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx
Comment (jerolimov) [inline frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx:19]: The `key` here is not wrong, but my first impression was: Oh it should be moved to the DetailsItem below..

But then I checked `frontend/public/components/default-resource.tsx` and saw that the loop adds already a key correctly to the usage of `<ExtensionDetailsItem`. So I think it's not needed here, and not on the DetailsItem below, or?

```suggestion
    <Component obj={obj} path={path} value={value} />
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
Thanks for applying the additional props to the `Component`. Looks almost good to me, I have just some tiny comments:

File: frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx
Comment (TheRealJon) [reply frontend/packages/console-shared/src/components/details-page/ExtensionDetailsItem.tsx:19]: Oh, yes, thanks for catching this. It was an oversight!

**Comment by TheRealJon** [comment]
@jerolimov I've addressed your comments. @rhamilto do you mind reviewing one last time?

**Review by rhamilto — APPROVED** [review]
/lgtm

**Review by jerolimov — APPROVED** [review]
re-apply lgtm after doc update

/lgtm

**Comment by yanpzhan** [comment]
@TheRealJon Hi Jon, I launched a cluster against the pr, checked normal details page for resource, eg console operator details page, and also deployed console-crontab-plugin dynamic plugin, checked crontab's details page. I wonder if there is a special dynamic plugin need to used to test the details-item? or do you know how to add one details-item from a dynamic plugin repo?

**Comment by yanpzhan** [comment]
@TheRealJon Tested from your branch[CONSOLE-3695-test](https://github.com/TheRealJon/console/tree/CONSOLE-3695-test) I could see the added detail items on console operator details page. The extension works now.
/label qe-approved

**Pull request merged**
